torchrec_dlrm/dlrm_main.py

```diff
@@ -439,7 +431,7 @@ def _train(
     combined_iterator = itertools.chain(
         iterator
         if limit_train_batches is None
-        else itertools.islice(iterator, limit_train_batches),
+        else itertools.islice(iterator, limit_train_batches - (epoch > 0)*(TRAIN_PIPELINE_STAGES - 1)),
```

Comment:
Regarding the change on line 434, let me explain.  

Epoch 0's first iteration in the training loop is the first time the forward pass occurs.  The first time a forward pass occurs, train_pipeline’s buffer is empty, so two additional batches are loaded.  Then the buffer is in the state it should stay in for the remainder of train_pipeline’s lifetime.

For epochs 1 and onward, train_pipeline’s buffer is also full.   Before epoch 1 runs its first iteration of the training loop, it should already have its first 2 batches loaded in the buffer.

Let me explain how this is relevant to the limit_train_batches variable.    When the buffer is not empty at the start of an epoch, that means the training iterator is already 2 steps through the training batch sequence.   If the sequence is to be limited by a number stored in limit_train_batches, then that number must be reduced by two to compensate for the iterator already being 2 steps in.  For example, suppose limit_train_batches = 5. For epoch 1, the iterator variable online 434 will be 2 batches into the training set batch sequence.  So to limit the sequence to 5 batches, that means only 3 more batches should be used.    But for epoch 0, the buffer is empty at the start, so the limit_train_batches number does not need to be altered to compensate 2 already-buffered batches because there are no 2 already-buffered batches for epoch 0.